File: src/main/java/net/greenjab/fixedminecraft/registry/block/NewIceBlock.java
package net.greenjab.fixedminecraft.registry.block;

import net.greenjab.fixedminecraft.registry.registries.GameruleRegistry;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.block.IceBlock;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.random.Random;
import net.minecraft.world.LightType;
import net.minecraft.world.attribute.EnvironmentAttributes;

public class NewIceBlock extends IceBlock {

    public NewIceBlock(Settings settings) {
        super(settings);
    }

    @Override
    protected void randomTick(BlockState state, ServerWorld world, BlockPos pos, Random random) {
        if (world.getGameRules().getValue(GameruleRegistry.Ice_Melt_In_Nether)) {
            if (world.getLightLevel(LightType.BLOCK, pos) > 11 - state.getOpacity() || world.getEnvironmentAttributes().getAttributeValue(EnvironmentAttributes.WATER_EVAPORATES_GAMEPLAY, pos))
                if (!nextToCryingObsidian(world, pos)) this.melt(state, world, pos);
        }
    }

    public static boolean nextToCryingObsidian(ServerWorld world, BlockPos pos) {
        for (Direction direction : Direction.values()) {
            if (world.getBlockState(pos.add(direction.getVector())).isOf(Blocks.CRYING_OBSIDIAN)) return true;
        }
        return false;
    }
}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/block/NewIceBlock.kt
package net.greenjab.fixedminecraft.registry.block

import net.greenjab.fixedminecraft.registry.GameruleRegistry
import net.minecraft.block.BlockState
import net.minecraft.block.IceBlock
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.random.Random
import net.minecraft.world.LightType

class NewIceBlock(settings: Settings) : IceBlock(settings) {
    @Suppress("OVERRIDE_DEPRECATION")
    override fun randomTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
        if (world.gameRules.getBoolean(GameruleRegistry.Ice_Melt_In_Nether)) {
            if (world.getLightLevel(LightType.BLOCK, pos) > 11 - state.getOpacity() || world.dimension.ultrawarm())
                this.melt(state, world, pos)
        }
    }
}
